### PR TITLE
ondemand_download_large_rel: solve flakyness

### LINKED ISSUE #3666: test_ondemand_download::test_ondemand_download_large_rel flakyness
Two ways:
- `assert 7 > 7` on the last line
- pageserver stop failed: pageserver with pid 18039 did not stop in 10 seconds [here](https://neon-github-public-dev.s3.amazonaws.com/reports/main/release/4235314752/index.html#suites/b97efae3a617afb71cb8142f5afa5224/da1f1dd229669df1/)

Noted, didn't fix yet:
- gc and compaction run on the second restart
- comments say that they are disabled, but period is non-zero, so they are executed

### PR DESCRIPTION
**Pull request by koivunej**
Disable background tasks to not get compaction downloading all layers but also stop safekeepers before checkpointing, use a readonly endpoint.

Fixes: #3666

### CODE REVIEW

**Comment by koivunej** [comment]
Not sure if this should actually fix this, since here it looks like the basebackup downloaded all of the layers, even if they were after the compaction? https://neon-github-public-dev.s3.amazonaws.com/reports/pr-3696/debug/4252834876/index.html#suites/b97efae3a617afb71cb8142f5afa5224/2486b335203eff3a/

Could it be that we cannot say that there's certainly some shuffling after this test workload where starting up the compute will require less than full amount of layers to be downloaded?

**Comment by koivunej** [comment]
Nevermind, probably misread that log, there's of course the initial size calculation which we are racing against as well, so earlier question rephrased:

Are we correct in expecting that there is a compaction result which will allow us to have less downloaded by basebackup+initial_size_calculation than the query?

Also, wondering if the compaction upload could go over timeout with real_s3, it will be racing against termination timeout -- not the 20s `wait_for_upload` because lsn doesn't change and `wait_for_upload` cannot see if compacted layers or image layers are still being uploaded. This effect has been hit on some other test case, probably `timetravel` where #3616 added a sleep.

**Comment by koivunej** [comment]
Rebased...

**Comment by koivunej** [comment]
Quite interesting *release* mode failure, seems like a lockup per pageserver logs: https://neon-github-public-dev.s3.amazonaws.com/reports/pr-3697/release/4353341505/index.html#suites/c19bc2126511ef8cb145cca25c438215/351c734cf7f081bd/?attachment=c4790db712f6e732

**Comment by koivunej** [comment]
> Also, wondering if the compaction upload could go over timeout with real_s3, it will be racing against 10s termination timeout -- not the 20s `wait_for_upload` because lsn doesn't change and `wait_for_upload` cannot see if compacted layers or image layers are still being uploaded.

I should probably try the `wait_for_upload_queue_empty` here added in #3741.

**Comment by koivunej** [comment]
This probably needs some feedback and ideas per my above comments.

File: test_runner/regress/test_ondemand_download.py [hunk 1]
```diff
@@ -98,6 +101,9 @@ def test_ondemand_download_large_rel(
     # wait until pageserver receives that data
     wait_for_last_record_lsn(client, tenant_id, timeline_id, current_lsn)
 
+    # stop pg before checkpoint to stop wal generation
+    pg.stop()
+
```
Comment (problame) [inline test_runner/regress/test_ondemand_download.py:106]: * shutdown pg
* wait_for_sk_commit_lsn_to_reach_remote_storage
* assert we have generated a bunch of layer files, e.g., using layermap endpoint
* start pageserver (not sks, not postgres)
* bring up read-only postgres (lsn=...) 
   * see https://github.com/neondatabase/neon/blob/a64dd3ecb58f02826d159dfe07a24b7ee52c9b82/test_runner/regress/test_disk_usage_eviction.py#L87-L95 for example
* do the query on the readonly postgres

absence of SKs means we don't have a logical size calculation from walreceiver.

also assert that the log doesn't contain any `logical_size.*download` messages ?

**Comment by koivunej** [comment]
Let's downscope the general failure to await for uploads in case we still have pending gc/compact out of this PR, rebased, should be good now.

**Review by LizardWizzard — APPROVED** [review]

File: test_runner/regress/test_ondemand_download.py (hunk 1 above)
Comment (problame) [reply test_runner/regress/test_ondemand_download.py]: Sorry for missing this.

The problem with plain `pg.stop()` (now `endpoint.stop()`) before the subsequent checkpoint is that the data written to pg might not yet have arrived at the PS when taking the checkpoint.

I think my suggestion to use `wait_for_sk_commit_lsn_to_reach_remote_storage` is obsolete, and the `wait_for_last_record_lsn` + moved `endpointstop()` is fine to ensure that the data written to PG reaches the pageserver before checkpoint.

What's still problematic is that the SKs are still running when we restart the pageserver.
That's what I was (badly) pointing out in
> absence of SKs means we don't have a logical size calculation from walreceiver.
> also assert that the log doesn't contain any logical_size.*download messages ?

Basically, if the SKs are still running upon PS restart, the PS is going to connect to the SK and it will calculate logical size for walreceiver status.
That means by the time we

```
before_downloads = get_num_downloaded_layers(client, tenant_id, timeline_id)
```

the on-demand downloads may have already happened and we'll trip

```
after_downloads = get_num_downloaded_layers(client, tenant_id, timeline_id)
log.info(f"layers downloaded before {before_downloads} and after {after_downloads}")
assert after_downloads > before_downloads
```

Hope this helps.

Comment (koivunej) [reply test_runner/regress/test_ondemand_download.py]: I'll put squash auto-merge optimistically, don't push resolve until you are happy with this.

**Review by problame — APPROVED** [review]
